src/cryptonote_core/cryptonote_core.cpp

```diff
@@ -1246,40 +1260,63 @@ namespace cryptonote
     }
 
     uint8_t version = m_blockchain_storage.get_current_hard_fork_version();
-    return m_mempool.add_tx(tx, tx_hash, blob, tx_weight, tvc, keeped_by_block, relayed, do_not_relay, version);
+    return m_mempool.add_tx(tx, tx_hash, blob, tx_weight, tvc, tx_relay, relayed, version);
   }
   //-----------------------------------------------------------------------------------------------
   bool core::relay_txpool_transactions()
   {
+    bool success = true;
+
     // we attempt to relay txes that should be relayed, but were not
-    std::vector<std::pair<crypto::hash, cryptonote::blobdata>> txs;
+    std::vector<std::tuple<crypto::hash, cryptonote::blobdata, relay_method>> txs;
     if (m_mempool.get_relayable_transactions(txs) && !txs.empty())
     {
-      cryptonote_connection_context fake_context = AUTO_VAL_INIT(fake_context);
-      tx_verification_context tvc = AUTO_VAL_INIT(tvc);
-      NOTIFY_NEW_TRANSACTIONS::request r;
-      for (auto it = txs.begin(); it != txs.end(); ++it)
+      NOTIFY_NEW_TRANSACTIONS::request public_req{};
+      NOTIFY_NEW_TRANSACTIONS::request private_req{};
+      for (auto& tx : txs)
       {
-        r.txs.push_back(it->second);
+        switch (std::get<2>(tx))
+        {
+          default:
+          case relay_method::none:
+          case relay_method::not_found:
+            break;
+          case relay_method::anonymity:
+            private_req.txs.push_back(std::move(std::get<1>(tx)));
+            break;
+          case relay_method::block:
+          case relay_method::flood:
+            public_req.txs.push_back(std::move(std::get<1>(tx)));
+            break;
+        }
       }
-      get_protocol()->relay_transactions(r, fake_context);
-      m_mempool.set_relayed(txs);
+
+      const boost::uuids::uuid source = boost::uuids::nil_uuid();
+      if (!public_req.txs.empty())
+        success &= get_protocol()->relay_transactions(public_req, source, epee::net_utils::zone::public_);
+      if (!private_req.txs.empty())
+        success &= get_protocol()->relay_transactions(private_req, source, epee::net_utils::zone::invalid);
```

Comment:
Good point. This is a pain in the butt to handle. I need to add a separate "timer" somewhere for this in core. It should probably be randomized too. Rats.

Reply:
I had to think a bit to re-work this one. The `send_tx` function now has a `delay` boolean to toggle a randomized delay. Its a bit of a hack, but it was the easiest to implement. The original leak would only occur if a public and "private" tx relay timer expired in the same 1 second interval.

Reply:
I scrapped everything I worked on, because it was pointless after Dandelion++ is added. So I decided to [PR diffusion / Dandelion++ fluff](https://github.com/monero-project/monero/pull/6146) immediately to fix this issue. This section of code will be updated with comments.